File: algo/src/main/java/com/hartwig/actin/algo/evaluation/molecular/GeneHasUTR3Loss.java
package com.hartwig.actin.algo.evaluation.molecular;

import java.util.Set;

import com.google.common.collect.Sets;
import com.hartwig.actin.PatientRecord;
import com.hartwig.actin.algo.datamodel.Evaluation;
import com.hartwig.actin.algo.datamodel.EvaluationResult;
import com.hartwig.actin.algo.evaluation.EvaluationFactory;
import com.hartwig.actin.algo.evaluation.EvaluationFunction;
import com.hartwig.actin.algo.evaluation.util.Format;
import com.hartwig.actin.molecular.datamodel.driver.CodingContext;
import com.hartwig.actin.molecular.datamodel.driver.Disruption;
import com.hartwig.actin.molecular.datamodel.driver.RegionType;
import com.hartwig.actin.molecular.datamodel.driver.Variant;
import com.hartwig.actin.molecular.datamodel.driver.VariantEffect;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class GeneHasUTR3Loss implements EvaluationFunction {

    @NotNull
    private final String gene;

    GeneHasUTR3Loss(@NotNull final String gene) {
        this.gene = gene;
    }

    @NotNull
    @Override
    public Evaluation evaluate(@NotNull PatientRecord record) {
        Set<String> hotspotsIn3UTR = Sets.newHashSet();
        Set<String> hotspotsIn3UTRUnreportable = Sets.newHashSet();
        Set<String> vusIn3UTR = Sets.newHashSet();

        for (Variant variant : record.molecular().drivers().variants()) {
            if (variant.gene().equals(gene) && variant.canonicalImpact().effects().contains(VariantEffect.THREE_PRIME_UTR)) {
                if (variant.isHotspot() && variant.isReportable()) {
                    hotspotsIn3UTR.add(variant.event());
                } else if (variant.isHotspot()) {
                    hotspotsIn3UTRUnreportable.add(variant.event());
                } else {
                    vusIn3UTR.add(variant.event());
                }
            }
        }

        Set<String> disruptionsIn3UTR = Sets.newHashSet();
        for (Disruption disruption : record.molecular().drivers().disruptions()) {
            if (disruption.gene().equals(gene) && disruption.codingContext() == CodingContext.UTR_3P
                    && disruption.regionType() == RegionType.EXONIC) {
                disruptionsIn3UTR.add(disruption.event());
            }
        }

        if (!hotspotsIn3UTR.isEmpty()) {
            return EvaluationFactory.unrecoverable()
                    .result(EvaluationResult.PASS)
                    .addAllInclusionMolecularEvents(hotspotsIn3UTR)
                    .addPassSpecificMessages(
                            "3' UTR hotspot mutation(s) in " + gene + " should lead to 3' UTR loss: " + Format.concat(hotspotsIn3UTR))
                    .addPassGeneralMessages("3' UTR loss of " + gene)
                    .build();
        }

        Evaluation potentialWarnEvaluation = evaluatePotentialWarns(hotspotsIn3UTRUnreportable, vusIn3UTR, disruptionsIn3UTR);

        if (potentialWarnEvaluation != null) {
            return potentialWarnEvaluation;
        }

        return EvaluationFactory.unrecoverable()
                .result(EvaluationResult.FAIL)
                .addFailSpecificMessages("No variants detected in 3' UTR region of " + gene)
                .addFailGeneralMessages("No 3' UTR loss of " + gene)
                .build();
    }

    @Nullable
    private Evaluation evaluatePotentialWarns(@NotNull Set<String> vusIn3UTR, @NotNull Set<String> hotspotsIn3UTRUnreportable,
            @NotNull Set<String> disruptionsIn3UTR) {
        Set<String> warnEvents = Sets.newHashSet();
        Set<String> warnSpecificMessages = Sets.newHashSet();
        Set<String> warnGeneralMessages = Sets.newHashSet();

        if (!hotspotsIn3UTRUnreportable.isEmpty()) {
            warnEvents.addAll(hotspotsIn3UTRUnreportable);
            warnSpecificMessages.add(
                    "Hotspot mutation detected in 3' UTR region of " + gene + " which may lead to 3' UTR loss: "
                            + Format.concat(hotspotsIn3UTRUnreportable) + " but variant is not considered reportable");
            warnGeneralMessages.add("Potential 3' UTR loss of " + gene);
        }

        if (!vusIn3UTR.isEmpty()) {
            warnEvents.addAll(vusIn3UTR);
            warnSpecificMessages.add(
                    "VUS mutation detected in 3' UTR region of " + gene + " which may lead to 3' UTR loss: " + Format.concat(vusIn3UTR));
            warnGeneralMessages.add("Potential 3' UTR loss of " + gene);
        }

        if (!disruptionsIn3UTR.isEmpty()) {
            warnEvents.addAll(disruptionsIn3UTR);
            warnSpecificMessages.add(
                    "Disruption(s) detected in 3' UTR region of " + gene + " which may lead to 3' UTR loss: " + Format.concat(
                            disruptionsIn3UTR));
            warnGeneralMessages.add("Potential 3' UTR loss of " + gene);
        }

        if (!warnEvents.isEmpty() && !warnSpecificMessages.isEmpty() && !warnGeneralMessages.isEmpty()) {
            return EvaluationFactory.unrecoverable()
                    .result(EvaluationResult.WARN)
                    .addAllInclusionMolecularEvents(warnEvents)
                    .addAllWarnSpecificMessages(warnSpecificMessages)
                    .addAllWarnGeneralMessages(warnGeneralMessages)
                    .build();
        }

        return null;
    }
}

File: algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/molecular/GeneHasUTR3Loss.kt
package com.hartwig.actin.algo.evaluation.molecular

import com.hartwig.actin.algo.evaluation.EvaluationLabels
import com.hartwig.actin.algo.evaluation.EvaluationFactory
import com.hartwig.actin.algo.evaluation.util.Format.concatVariants
import com.hartwig.actin.datamodel.algo.Evaluation
import com.hartwig.actin.datamodel.molecular.MolecularTest
import com.hartwig.actin.datamodel.molecular.MolecularTestTarget
import com.hartwig.actin.datamodel.molecular.driver.CodingContext
import com.hartwig.actin.datamodel.molecular.driver.Disruption
import com.hartwig.actin.datamodel.molecular.driver.RegionType
import com.hartwig.actin.datamodel.molecular.driver.VariantEffect

class GeneHasUTR3Loss(override val gene: String, labels: EvaluationLabels.Molecular): MolecularEvaluationFunction(
    targetCoveragePredicate = specific(MolecularTestTarget.MUTATION, labels.geneHasUtr3LossMessagePrefix()),
    labels = labels
) {

    override fun evaluate(test: MolecularTest): Evaluation {
        val (cavsIn3UTR, cavsIn3UTRUnreportable, vusIn3UTR) = test.drivers.variants.filter { variant ->
            variant.gene == gene && variant.canonicalImpact.effects.contains(VariantEffect.THREE_PRIME_UTR)
        }
            .fold(Triple(emptySet<String>(), emptySet<String>(), emptySet<String>())) { acc, variant ->
                if (variant.isCancerAssociatedVariant && variant.isReportable) {
                    acc.copy(first = acc.first + variant.event)
                } else if (variant.isCancerAssociatedVariant) {
                    acc.copy(second = acc.second + variant.event)
                } else {
                    acc.copy(third = acc.third + variant.event)
                }
            }

        val disruptionsIn3UTR = test.drivers.disruptions.filter { disruption ->
            disruption.gene == gene && disruption.codingContext == CodingContext.UTR_3P && disruption.regionType == RegionType.EXONIC
        }
            .map(Disruption::event)
            .toSet()

        if (cavsIn3UTR.isNotEmpty()) {
            return EvaluationFactory.pass(
                labels.geneHasUtr3LossPass(concatVariants(cavsIn3UTR, gene), gene),
                inclusionEvents = cavsIn3UTR
            )
        }
        val potentialWarnEvaluation = evaluatePotentialWarns(cavsIn3UTRUnreportable, vusIn3UTR, disruptionsIn3UTR)
        return potentialWarnEvaluation ?: EvaluationFactory.fail(labels.geneHasUtr3LossFail(gene))
    }

    private fun evaluatePotentialWarns(
        vusIn3UTR: Set<String>, cavsIn3UTRUnreportable: Set<String>, disruptionsIn3UTR: Set<String>
    ): Evaluation? {
        return MolecularEventUtil.evaluatePotentialWarnsForEventGroups(
            listOf(
                EventsWithMessages(
                    cavsIn3UTRUnreportable,
                    labels.geneHasUtr3LossWarnUnreportable(concatVariants(cavsIn3UTRUnreportable, gene), gene)
                ),
                EventsWithMessages(
                    vusIn3UTR,
                    labels.geneHasUtr3LossWarnVus(concatVariants(vusIn3UTR, gene), gene)
                ),
                EventsWithMessages(
                    disruptionsIn3UTR,
                    labels.geneHasUtr3LossWarnDisruption(concatVariants(disruptionsIn3UTR, gene), gene)
                )
            )
        )
    }
}